modules/rostests/apitests/kernel32/GetFinalPathNameByHandle.c

```diff
@@ -0,0 +1,464 @@
+    ok_eq_ulong(Result, ExpectedStringLength + 1);
+    ok_eq_wchar(Buffer[0], 0xCCCC);
+    ok_err(ERROR_NOT_ENOUGH_MEMORY);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, 9, VOLUME_NAME_DOS);
+    ok_eq_ulong(Result, ExpectedStringLength + 1);
+    ok_eq_wchar(Buffer[0], 0xCCCC);
+    ok_err(ERROR_NOT_ENOUGH_MEMORY);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, NULL, 0, VOLUME_NAME_DOS);
+    ok_eq_ulong(Result, ExpectedStringLength + 1);
+    ok_err(ERROR_NOT_ENOUGH_MEMORY);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), 0x1000);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), 7);
+    ok_eq_ulong(Result, 0ul);
+    ok_eq_wchar(Buffer[0], 0xCCCC);
+    ok_err(ERROR_INVALID_PARAMETER);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_DOS);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_DOS | FILE_NAME_OPENED);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    /* Query the GUID based volume name */
+    memcpy(VolumeBuffer, FilePath, 3 * sizeof(WCHAR));
+    VolumeBuffer[3] = UNICODE_NULL;
+    Success = GetVolumeNameForVolumeMountPointW(VolumeBuffer,
+                                                ExpectedString,
+                                                ARRAYSIZE(ExpectedString));
+    if (!Success)
+    {
+        skip("GetVolumeNameForVolumeMountPointW failed: %ld\n", GetLastError());
+        return;
+    }
+
+    /* Expected string for VOLUME_NAME_GUID:
+       L"\\\\?\\Volume{cd4317d4-A62f-53d7-b36c-73f935c37280}\\ReactOS\\bin\\kernel32_apitest.exe" */
+    VolumeRelativePath = &FilePath[wcslen(L"C:\\")];
+    wcscat(ExpectedString, VolumeRelativePath);
+    ExpectedStringLength = wcslen(ExpectedString);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID);
+    ok_eq_ulong(Result, ExpectedStringLength);
+    ok_eq_wstr(Buffer, ExpectedString);
+    ok_err(0);
+
+    SetLastError(0xdeadbeef);
+    memset(Buffer, 0xCC, sizeof(Buffer));
+    Result = pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
```

Comment:
Your implementation doesn't seem to do anything with either `FILE_NAME_NORMALIZED` or `FILE_NAME_OPENED`.
Any reason why so (missing feature in the kernel or in kernel32/kernelbase, ...)?

Reply:
Those don't do anything that was visible. Feel free to write a test that shows that they do something.